src/words/ast/INodeListenerTemp.java

```diff
@@ -13,16 +13,25 @@ public ASTValue eval(WordsEnvironment environment) throws WordsRuntimeException
 		AST predicate = children.get(0);
 		AST statementList = children.get(1);
 		
-		ASTValue predicateValue = predicate.eval(environment);
-
-		// currently only restricted to boolean predicate
-		assert predicateValue.type == ASTValue.ValueType.BOOLEAN : "Predicate has type " + predicateValue.type.toString();
-
-		// temporary listener is created only if the predicate is true
-		if (predicateValue.booleanValue == true) {
-			environment.createListener(predicate, statementList, true);
+		if (predicate instanceof INodeBasicActionPredicate) {
+			ASTValue predicateValue = predicate.eval(environment, statementList);
+			
+			if (predicateValue.booleanValue == true) {
+				environment.createListener(predicate, statementList, true);
```

Comment:
I think there's a subtlety here that perhaps is unclear in the language specification.  Do we actually want to execute the statement list right now during this check?  Or just check that the predicate is true, but not evaluate the statement list now (wait until Phase 3 of the frame instead).  I'm concerned we might effectively execute the statement list twice in the same frame.